### ISSUE TITLE
How to Access Backbone Features and Add Custom DG Loss While Using Ultralytics Training Pipeline?

### ISSUE THREAD

**Issue by Kaustubh-Atey** [ultralytics/ultralytics#22674]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi, I want to add a domain generalization (DG) feature-level loss to YOLO training while still using all inbuilt Ultralytics components (optimizer, scheduler, EMA, augmentations, AMP, etc.).

**Goal:** Use a training loop conceptually like:
```
for img, labels, domain_info in dataloader:
    outputs, features = model(img)  # need backbone features
    f1, f2 = split_by_domain(features, domain_info)
    dg_loss = DGLoss(f1, f2)
    total_loss = yolo_loss(outputs, labels) + dg_loss
    total_loss.backward()
```

**What I need:**

1. How to access backbone features during training?
Is there a recommended way to modify the model forward pass (or use hooks) so it can return features without breaking the Ultralytics training pipeline?

2. Where to add a custom loss?
Is extending the Trainer class the right way, or should I use callbacks for injecting the DG loss?

3. Any examples?
If there are examples of returning (preds, features) or adding custom losses while retaining default training, that would help.

**Use case:** I’m training on multi-domain data and need to compute a DG loss between domain-specific backbone features.

Thanks for the guidance!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Kaustubh-Atey, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started quickly. An Ultralytics engineer will also review and assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including:
-Your current Ultralytics version and environment details
-A brief description of what you’ve tried so far (e.g., forward changes, hooks, callbacks, or Trainer subclassing)
-Any relevant code snippets or a small prototype showing where you plan to compute DG loss
-Dataset/image examples and training logs
Also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You would very likely need to modify the source code. You will need to inherit from `DetectionModel` and override the `loss` method similar to how it's done here:

https://github.com/ultralytics/ultralytics/blob/2d633d0fa1b80a25cfef9f0ba18aa6bdbc7b950e/ultralytics/nn/tasks.py#L739
